cocotb/scheduler.py

```diff
@@ -1072,6 +1074,12 @@ def _cleanup(self):
                     self.log.debug("Killing %s" % str(coro))
                 coro.kill()
 
+        # if there are coroutines being scheduled when the test ends, kill them (gh-1347)
```

Comment:
It's not immediately clear to me that `_cleanup` should be allowed to run while in the middle of the `_event_loop`.

Reply:
Intuitively, you are right. Unfortunately, it doesn't work that way. Maybe one day, after a significant refactor, the scheduler will be a little more intuitive. I've already tried a few times to do this without a lot of success.

If a task fails it is `_unschedule`d. If there are no waiters on it, then it `_abort_test`s. The test is then `_unschedule`d which results in a call to `_cleanup`.

Reply:
At some point we should probably have some PlantUML state diagrams or something like that in the docs.